### DISCUSSION TITLE
What do I do when the CI is failing?

### DISCUSSION THREAD

**Discussion by t-kimber** [volkamerlab/teachopencadd#187] [How Tos]
Is the CI on your new talktorial not passing the CI tests?

Here's a list of things you can try:
Assuming you are in the `teachopencadd` folder:

 - For the black test, run the following: `black-nb -l 99 teachopencadd/`
 - For the README test, run the following:
```
for path in teachopencadd/talktorials/T*/talktorial.ipynb; do
    python devtools/regenerate_readmes.py --output README.md $path
done
```
- For Python versions: you could ignore timeout errors. 
- Docs: mmhhh.. read the errors.

**Comment by AndreaVolkamer** [comment]
Addition for Pytest, in case you want to check the behavior of one specific notebook locally on your own system at least, you could try the following:
* specify the required python version in the yml file 
   * `devtools/test_env.yml`
   *  adapt line 6, e.g. `python=3.7`
* create a new test environment
   * `mamba env create -f devtools/test_env.yml -n toc-ci`
* activate it
   *  `conda activate toc-ci` 
* set pytest arguments
   *  `PYTEST_ARGS="--nbval-lax --current-env --dist loadscope --numprocesses 2"`
* run pytest with the talktorial in question (here T018)
   *  `pytest $PYTEST_ARGS teachopencadd/talktorials/T018_automated_cadd_pipeline/talktorial.ipynb`

**Comment by dominiquesydow** [comment]
I was trying to use `act` to run the GHA workflows locally (in my case on MacOS, M1 chip):

- Follow installation instructions here: https://github.com/nektos/act#installation-through-package-managers
- We are using `miniconda` in our workflows, which is not actually installed during the run and instead searches for such an installation during the run, which seems to be available on GHA servers but not on my computer. Leaving here a potential fix once I pick this up again: https://github.com/lmmx/devnotes/wiki/Running-GitHub-CI-locally-with-nektos-act-(Docker)-in-LXC

Once I or anyone else gets this set up, let's write up a proper documentation on "How to run TeachOpenCADD workflows locally via `act`".

More notes:
- Currently, `act` supports only Linux runners (i.e. we cannot run our MacOS and Windows CIs): https://github.com/nektos/act#runners
